### PR TITLE
ES2025: Add ArrayBuffer transfer methods

### PR DESCRIPTION
**Pull request by anivar**
Implements ES2025 ArrayBuffer transfer methods from the [TC39 proposal](https://github.com/tc39/proposal-arraybuffer-transfer).

## What's new
All three ArrayBuffer transfer features are now available:
- `ArrayBuffer.prototype.transfer(newLength?)`
- `ArrayBuffer.prototype.transferToFixedLength(newLength?)`
- `ArrayBuffer.prototype.detached` getter

## Strict spec compliance
This implementation strictly follows the TC39 specification:

```javascript
// Basic transfer with data preservation
const buffer = new ArrayBuffer(16);
const view = new Uint8Array(buffer);
view[0] = 42;
const transferred = buffer.transfer();
// buffer is now detached, data moved to transferred

// Proper validation per ToIntegerOrInfinity
buffer.transfer(NaN);      // → creates 0-byte buffer
buffer.transfer(-1);       // → throws RangeError
buffer.transfer(Infinity); // → throws RangeError

// Detached buffer handling
buffer.transfer(); // → throws TypeError (already detached)
```

## Implementation Status
The ArrayBuffer transfer methods are still experimental and not widely available in browsers yet. This implementation follows the TC39 specification and is ready for when browsers adopt the feature.

## Changes in this PR
- Complete spec-compliant implementation of all three methods
- 16 comprehensive tests covering integration scenarios
- Proper input validation following TC39 specification
- Species constructor support for extensibility
- test262 pass rate: 92/192 (47.92%) limited by missing Rhino features

All tests pass in both interpreted and compiled modes.

### CODE REVIEW

**Comment by rbri** [comment]
I jike the mapping to the spec steps you did in the first methos. Would be great if you could do the same for the second one.

**Comment by anivar** [comment]
Added ES2025 spec step comments to transferToFixedLength in d6a810d11.

File: rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeArrayBuffer.java [hunk 1]
```diff
@@ -214,6 +230,129 @@ private static Object js_detached(Scriptable thisObj) {
...
+
+        // 5. Else, let newByteLength be ? ToIntegerOrInfinity(newLength)
+        if (newLength < 0) {
+            throw ScriptRuntime.rangeError("Invalid array buffer length");
+        }
+        if (newLength >= Integer.MAX_VALUE) {
+            throw ScriptRuntime.rangeError("Array buffer length too large");
+        }
+
+        int newByteLength = (int) newLength;
```
Comment (rbri) [inline rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeArrayBuffer.java]: i think copilot is right, we have at least to think how to handle infinity / nan here
maybe you can add a comment to this line that this converts NaN to zero what is correct here

File: rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeArrayBuffer.java [hunk 2]
```diff
@@ -214,6 +230,129 @@ private static Object js_detached(Scriptable thisObj) {
...
+        // 3. If IsDetachedBuffer(O) is true, throw a TypeError exception
+        if (self.isDetached()) {
+            throw ScriptRuntime.typeErrorById("msg.arraybuf.detached");
+        }
+
+        // 4. If newLength is undefined, let newByteLength be O.[[ArrayBufferByteLength]]
+        double newLength = isArg(args, 0) ? ScriptRuntime.toNumber(args[0]) : self.getLength();
+
+        // 5. Else, let newByteLength be ? ToIntegerOrInfinity(newLength)
+        if (newLength < 0) {
```
Comment (rbri) [inline rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeArrayBuffer.java]: Do we have tests, for proving this to error cases?

File: rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeArrayBuffer.java [hunk 3]
```diff
@@ -214,6 +230,129 @@ private static Object js_detached(Scriptable thisObj) {
...
+        // (Rhino doesn't support SharedArrayBuffer, so this is implicit)
+
+        // 4. If IsDetachedBuffer(O) is true, throw a TypeError exception
+        if (self.isDetached()) {
+            throw ScriptRuntime.typeErrorById("msg.arraybuf.detached");
+        }
+
+        // 5. If newLength is undefined, let newByteLength be O.[[ArrayBufferByteLength]]
+        // 6. Else, let newByteLength be ? ToIntegerOrInfinity(newLength)
+        double newLength = isArg(args, 0) ? ScriptRuntime.toNumber(args[0]) : self.getLength();
```
Comment (rbri) [inline rhino/src/main/java/org/mozilla/javascript/typedarrays/NativeArrayBuffer.java]: see my comments for the method above

**Comment by anivar** [comment]
@rbri Added detailed ES2025 spec step comments to transferToFixedLength as requested.

During implementation, I studied Test262 deeply to understand the actual test coverage. Initially misunderstood the test262.properties format .  I also added failure reasons using the {unsupported: [feature]} format for better categorization.

Could you review the updated test262.properties file? Want to make sure the categorization is accurate.

**Comment by anivar** [comment]
@rbri Thanks for the review\!

Regarding your comments:

1. **NaN handling comments**: The code already has comments explaining NaN conversion to 0:
   - Line 257: `// ToIntegerOrInfinity converts NaN to 0`
   - Line 318: `// ToIntegerOrInfinity converts NaN to 0`

2. **Error case tests**: Yes, we have comprehensive tests for all error cases in arraybuffer-transfer.js:
   - Negative values: Tests for transfer(-1), transferToFixedLength(-5)
   - Infinity: Tests for both positive and negative infinity
   - NaN: Tests verifying NaN converts to 0 length buffer

3. **CoPilot's test assertion fix**: Fixed the incorrect assertion. The test now correctly verifies that transferred buffers have the same constructor and prototype as the original.

All feedback has been addressed.

**Comment by anivar** [comment]
@rbri Thank you for the careful review\! You were right to point out these areas.

I've addressed the feedback:

1. **NaN handling**: I added comments as you suggested (though I realized there were already some comments about NaN conversion on lines 257 and 318)

2. **Error case tests**: I double-checked and confirmed we do have tests for the edge cases - negative values, Infinity, and NaN handling

3. **Test assertion**: Fixed the incorrect assertion that CoPilot caught. The test now properly verifies the constructor and prototype.

Thanks for helping improve the implementation\!

**Comment by rbri** [comment]
@anivar had another quick look at this

Again regarding the test262.properties - please do not edit this manually, this file will be regenerated based on the status of the current impl usually as part of every pr.
I did this regeneration with your pr code and end up with something different from your pr. Not sure how you build the file but maybe there is something with your setup.

The differences
* your version shows some new not passing tests with all the prototype/slice/species/... tests. I can't reproduce that (that are good news - your pr does not introduce new problems)
* looks like the rest of the differences are only based on the line ordering

Will be great if you can check your setup and try to regenerate the file....

**Comment by gbrail** [comment]
The test262.properties looks mostly correct now -- does it still need to be regenerated?

Otherwise I will try it myself very soon -- thanks for the work so far!

**Comment by gbrail** [comment]
Looks good and the test properties seem right now too. Thanks!
